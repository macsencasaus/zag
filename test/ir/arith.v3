// RUN: %vector -emit-ir %s

fn square(x: i32): i32 {
    return x * x;
}

fn sum(a: i32, b: i32, c: i32): i32 {
    return a + b + c;
}

fn prec(): i64 {
    var x: i64 = 1 + 2 * 3;
    var y: i64 = 1 / 2 - 3;
    return x % y;
}

fn mismatch(): i32 {
    var y: i32 = 3;
    return y + 2;
}

fn prec2(): i64 {
    return 2 * 2 + 3 * 3;
}

fn prec3(): i64 {
    return 2 * (2 + 3) * 3;
}

fn prefix(): i32 {
    var x: i32 = 2, y: i32;
    x = -x;
    x = !x;
    x = ~x;
    ++x;
    y = x++;
    --x;
    y = x--;
    return x;
}

fn more_binops(): i32 {
    var x: i32, y: i32;
    y = x < 1;
    y = x <= 2;
    y = x == 3;
    y = x != y;
    y = x > y;
    y = x >= y;
    y = x & y;
    y = x | y;
    y = x ^ x;
    y = x << 1;
    y = x >> 1;
    return 0;
}
