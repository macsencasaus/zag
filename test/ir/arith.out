; V3 Readable Intermediate Representation
; Function header formatted as:
; @func[<max-stack-size>](<param-type>,...)

@square[8](i32):
    %s[4] = imul i32 %s[0], i32 %s[0]
    ret i32 %s[4]

@sum[20](i32, i32, i32):
    %s[12] = add i32 %s[0], i32 %s[4]
    %s[16] = add i32 %s[12], i32 %s[8]
    ret i32 %s[16]

@prec[24]():
    %s[0] = imul i64 2, i64 3
    %s[8] = add i64 1, i64 %s[0]
    %s[0] = i64 %s[8]
    %s[8] = idiv i64 1, i64 2
    %s[16] = sub i64 %s[8], i64 3
    %s[8] = i64 %s[16]
    %s[16] = imod i64 %s[0], i64 %s[8]
    ret i64 %s[16]

@mismatch[8]():
    %s[0] = i32 3
    %s[4] = add i32 %s[0], i32 2
    ret i32 %s[4]

@prec2[24]():
    %s[0] = imul i64 2, i64 2
    %s[8] = imul i64 3, i64 3
    %s[16] = add i64 %s[0], i64 %s[8]
    ret i64 %s[16]

@prec3[24]():
    %s[0] = add i64 2, i64 3
    %s[8] = imul i64 2, i64 %s[0]
    %s[16] = imul i64 %s[8], i64 3
    ret i64 %s[16]

@prefix[12]():
    %s[0] = i32 2
    %s[8] = neg i32 %s[0]
    %s[0] = i32 %s[8]
    %s[8] = lnot i32 %s[0]
    %s[0] = i32 %s[8]
    %s[8] = bnot i32 %s[0]
    %s[0] = i32 %s[8]
    %s[0] = add i32 %s[0], i32 1
    %s[8] = i32 %s[0]
    %s[0] = add i32 %s[0], i32 1
    %s[4] = i32 %s[8]
    %s[0] = sub i32 %s[0], i32 1
    %s[8] = i32 %s[0]
    %s[0] = sub i32 %s[0], i32 1
    %s[4] = i32 %s[8]
    ret i32 %s[0]

@more_binops[12]():
    %s[8] = slt i32 %s[0], i32 1
    %s[4] = i32 %s[8]
    %s[8] = sle i32 %s[0], i32 2
    %s[4] = i32 %s[8]
    %s[8] = eq i32 %s[0], i32 3
    %s[4] = i32 %s[8]
    %s[8] = ne i32 %s[0], i32 %s[4]
    %s[4] = i32 %s[8]
    %s[8] = sgt i32 %s[0], i32 %s[4]
    %s[4] = i32 %s[8]
    %s[8] = sge i32 %s[0], i32 %s[4]
    %s[4] = i32 %s[8]
    %s[8] = and i32 %s[0], i32 %s[4]
    %s[4] = i32 %s[8]
    %s[8] = or i32 %s[0], i32 %s[4]
    %s[4] = i32 %s[8]
    %s[8] = xor i32 %s[0], i32 %s[0]
    %s[4] = i32 %s[8]
    %s[8] = shl i32 %s[0], i32 1
    %s[4] = i32 %s[8]
    %s[8] = ashr i32 %s[0], i32 1
    %s[4] = i32 %s[8]
    ret i32 0

